src/passes/MinimizeRecGroups.cpp

```diff
@@ -0,0 +1,752 @@
+      classInfo.advance();
+      classInfo.permute(groupInfo);
+
+      shapesToUpdate.push_back(group);
+      return;
+    }
+
+    // Case 2B: We need to join two matching unaffiliated groups into a new
+    // nontrivial equivalence class.
+    assert(!groups[groupRep].classInfo && !groups[otherRep].classInfo);
+    assert(group == groupRep && other == otherRep);
+
+    // We are canonicalizing and reinserting the shape for other, so remove it
+    // from the map under its current shape.
+    groupShapeIndices.erase(it);
+
+    // Put the types for both groups in to the canonical order.
+    auto permutation = getCanonicalPermutation(groupInfo.group);
+    groupInfo.permutation = otherInfo.permutation = std::move(permutation);
+
+    // Set up `other` to be the representative element of the equivalence class.
+    otherInfo.classInfo.emplace(otherInfo);
+    auto& classInfo = *otherInfo.classInfo;
+
+    // Update both groups to have the initial valid order. Their shapes still
+    // match.
+    classInfo.permute(otherInfo);
+    classInfo.permute(groupInfo);
+
+    // Insert `other` with its new shape. It may end up being joined to another
+    // existing equivalence class, in which case its class info will be cleared
+    // and `group` will subsequently be added to that same existing class, or
+    // alternatively it may be inserted as the representative element of a new
+    // class to which `group` will subsequently be joined.
+    shapesToUpdate.push_back(group);
+    shapesToUpdate.push_back(other);
+  }
+
+  std::vector<size_t>
+  getCanonicalPermutation(const std::vector<HeapType>& types) {
+    // The correctness of this part depends on some interesting properties of
+    // strongly connected graphs with ordered, directed edges. A permutation of
+    // the vertices in a graph is an isomorphism that produces an isomorphic
+    // graph. An automorphism is an isomorphism of a structure onto itself, so
+    // an automorphism of a graph is a permutation of the vertices that does not
+    // change the graph. Permutations can be described in terms of sets of
+    // cycles of elements.
+    //
+    // Theorem 1: All cycles in an automorphism of an SCC are the same size.
+    //
+    //     Proof: By contradiction. Assume there are two cycles of different
+    //     sizes. Because the SCC is fully connected, there is a path from an
+    //     element in the smaller of these cycles to an element in the larger.
+    //     This path must contain an edge between two edges in cycles of
+    //     different sizes N and M such that N < M. Apply the automorphism N
+    //     times. The source of this edge has been cycled back to its original
+    //     index, but the destination is not yet back at its original index, so
+    //     the edge's destination index is different from its original
+    //     destination index and the permutation we applied was not an
+    //     automorphism after all.
```

Comment:
I have yet to fully understand this, but it sticks out to me that I don't see how you are using the **auto**morphism property (that is, that rec group shape rules indicate equality). Rather it seems like you are only using the **iso**morphism property, of having a permutation. Can this theorem not be stated on cycles in a permutation on an SCC?

Reply:
An automorphism is an isomorphism, so automorphisms are permutations, too.

Here is an interesting rec group:

```
(rec 
  (type $a1 (struct (field (ref $a2) (ref $b1))))
  (type $a2 (struct (field (ref $a1) (ref $b2))))
  (type $b1 (struct (field (ref $b2) (ref $a1))))
  (type $b2 (struct (field (ref $b1) (ref $a2))))
)
```

In [cycle notation](https://en.wikipedia.org/wiki/Permutation#Cycle_notation), here is an automorphism of this rec group: `($a1 $a2)($b1 $b2)`. Note that the cycles in the automorphism are the same size (2). This theorem states that it is impossible for there to be an automorphism of any SCC of a type definition graph in which the cycles are not all the same size.

Reply:
Thanks for the example, but I think I wasn't clear enough in my question, sorry.

> An automorphism is an isomorphism, so automorphisms are permutations, too.

Yes, but then the theorem could be stated on permutations, which would be better? That is, when I see a theorem in geometry that begins

1. Let there be a triangle ABC
2. Let the angle ACB be 55 degrees
3. ..

Then I will skim the proof to see how fact 2 is used. If it is never used then the theorem does not need it, and should not mention it.

Again, my question is: how are the rec group canonicalization rules used in this proof? I can't find that. And those rules should be used because our previous discussion ended with the conclusion that what is new in an automorphism (compared to an isomorphism) is that it maps something that by those rules is equal to itself.

If this theorem is valid for any isomorphism, then it could be written that way, and a sentence after the proof ends could be "As a quick corollary, any automorphism is an isomorphism, so the above applies to automorphisms as well." But I think you must be using the rules somehow, I just can't see where. Perhaps that is because I am stuck on the next part:

> This path must contain an edge between two edges

I'm not sure how to read this - is it a typo? (edges typically don't connect edges). I guess this might be using "edge" in two senses, edges of the SCC and edges of the permutation - if so, please specify which is which.

Reply:
This theorem is *not* true for any isomorphism! Here's a perfectly good isomorphism that does not have cycles of equal size: `($a1 $a2 $b1) ($b2)`. The fact that we are looking at an automorphism is critical because without that, we don't have a contradiction after applying the automorphism N times.

> This path must contain an edge between two edges

Typo. Should be "this path must contain an edge between two vertices."

Reply:
Offline discussion clarified to me how rec group canonicalization works (in my brain I had some legacy notions from the equirecursive days). After that + the typo fix, this proof checks out for me.